fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBPreloadRecordCache.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * FDBRecordCache.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2020 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.provider.foundationdb;
+
+import com.apple.foundationdb.annotation.API;
+import com.apple.foundationdb.tuple.Tuple;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+/**
+ * A cache used to track and handle the results of asynchronously preloaded records.
+ */
+@API(API.Status.INTERNAL)
+class FDBPreloadRecordCache {
+    @Nonnull
+    private final Cache<Tuple, EntryImpl>  cache;
+
+    public FDBPreloadRecordCache(int maximumSize) {
+        cache = CacheBuilder.newBuilder().maximumSize(maximumSize).build();
+    }
+
+    /**
+     * Called prior to performing a prefetch for a given {@code Tuple}, returning a future holder that must
+     * be filled in with the record when it has completed.
```

Comment:
What is the purpose of having incomplete entries in the cache at all? They aren't returned to another load attempt or even shared by it?

Reply:
I should have commented on this.  The current strategy allows for a race condition I believe. 

```
Thread 1                    Thread 2
-----------------      ------------------
begin pre-fetch    
read record
                       replace record
                       invalidate (non-existant) cache entry
put record in cache
```

with the strategy I'm proposing, the holder for the record being pre-fetched is pre-allocated in the cache and completed when the read is finished.  An invalidate kicks the holder out of the cache, so that the results will never be seen.

Reply:
There *may* even be a path for the pre-fetch read to pick up a portion of the replaced record as well....

Reply:
Okay. I guess there should be a short comment there for why it does that.

Such race conditions seem more fundamentally problematic. For instance,
```
Thread 1                    Thread 2
-----------------      ------------------
begin pre-fetch    
read record
                       replace record
                       invalidate (incomplete) cache entry
(put record in orphan entry)
take action based on out-of-date record
```
But there's no reason not to try to avoid what can be relatively cheaply.

Reply:
I *think* we are saying the same thing...unless I'm interpreting your example wrong.  Since the orphaned entry is ejected from the cache, its value (the record that thread #1 was reading), will never be visible to the outside world to act upon it.   But, yes, I'll add a comment to explain the reason for the design.

Reply:
Yes, but thread 1 continues to make changes elsewhere dependent on stale information. I am not saying we need to try to fix this here. Or that we should give up on concurrency within the same context because of it. Just that it's hard to solve without synchronization outside of what we provide.

Reply:
Got it.